### ISSUE TITLE
Message stanza can be stored with any value for the "type" attribute.

### ISSUE THREAD

**Issue by iAchilles** [processone/ejabberd#1350]
Hello! 
From [RFC-6121](https://tools.ietf.org/html/rfc6121#section-5.2.2): 

> Inclusion of the 'type' attribute is RECOMMENDED.
> If included, the 'type' attribute MUST have one of the following values:
> - chat 
> - error 
> - groupchat 
> - headline 
> - normal 

But it is possible now to send the following stanza

```
<message to='alice@domain.com'  type='abcdefghij'>
<body>Hey!</body>
</message>
```

This stanza will be stored in the 'archive' table as is. 
In theory, a client can generate any values for the "type" attribute. 
Also, I noticed that `mod_mam_sql:select/8` uses `jlib:binary_to_atom/1` to convert 'type' value to atom.
Is it safe? 

P.S. Since v16.09 I can use a filter in the `store_mam_message` hook to control "type" attribute, but why is not implemented by default?

**Comment by weiss** [comment]
> Since v19.09 I can use a filter in the `store_mam_message` hook to control "type" attribute, but why is not implemented by default?

Because RFC-6121 also [says][1]:

> If an application receives a message with no 'type' attribute or the application does not understand the value of the 'type' attribute provided, it MUST consider the message to be of type "normal" (i.e., "normal" is the default).

> Also, I noticed that `mod_mam_sql:select/8` uses `jlib:binary_to_atom/1` to convert 'type' value to atom.
Is it safe?

Unless I'm overlooking something, the `Kind` [handed over][2] to `jlib:binary_to_atom/1` is guaranteed to be either [`chat`][3] or [`groupchat`][4].  So yes, as far as I can see, it's safe.

Did you run into an actual problem you're trying to solve?

[1]: https://tools.ietf.org/html/rfc6121#section-5.2.2
[2]: https://github.com/processone/ejabberd/blob/ed6bc9081ba0544b384a9fefc33b0599a3f790f3/src/mod_mam_sql.erl#L168
[3]: https://github.com/processone/ejabberd/blob/ed6bc9081ba0544b384a9fefc33b0599a3f790f3/src/mod_mam.erl#L727
[4]: https://github.com/processone/ejabberd/blob/ed6bc9081ba0544b384a9fefc33b0599a3f790f3/src/mod_mam.erl#L739

**Comment by zinid** [comment]
This issue is fixed in `xml-ng` branch, which will be merged in 16.11.

**Comment by zinid** [comment]
The issue with unknown stanza type is [fixed](https://github.com/processone/xmpp/commit/197a372da160bbade3dc40f67b6e51d6ac89832d) in xmpp library.
The fix will be included in ejabberd 17.01.
